File: aria2c/src/main/java/com/yausername/aria2c/Aria2c.java
package com.yausername.aria2c;

import android.content.Context;

import androidx.annotation.NonNull;

import com.yausername.youtubedl_android.YoutubeDLException;
import com.yausername.youtubedl_common.SharedPrefsHelper;
import com.yausername.youtubedl_common.utils.ZipUtils;

import org.apache.commons.io.FileUtils;

import java.io.File;

public class Aria2c {

    private static final Aria2c INSTANCE = new Aria2c();
    protected static final String baseName = "youtubedl-android";
    private static final String packagesRoot = "packages";
    private static final String aria2cDirName = "aria2c";
    private static final String aria2cLibName = "libaria2c.zip.so";
    private static final String aria2cLibVersion = "aria2cLibVersion";

    private boolean initialized = false;
    private File binDir;

    private Aria2c() {
    }

    public static Aria2c getInstance() {
        return INSTANCE;
    }

    synchronized public void init(Context appContext) throws YoutubeDLException {
        if (initialized) return;

        final File baseDir = new File(appContext.getNoBackupFilesDir(), baseName);
        if (!baseDir.exists()) baseDir.mkdir();

        binDir = new File(appContext.getApplicationInfo().nativeLibraryDir);

        final File packagesDir = new File(baseDir, packagesRoot);
        final File aria2cDir = new File(packagesDir, aria2cDirName);
        initAria2c(appContext, aria2cDir);

        initialized = true;
    }

    private void initAria2c(Context appContext, File aria2cDir) throws YoutubeDLException {
        final File aria2cLib = new File(binDir, aria2cLibName);
        // using size of lib as version
        final String aria2cSize = String.valueOf(aria2cLib.length());
        if (!aria2cDir.exists() || shouldUpdateAria2c(appContext, aria2cSize)) {
            FileUtils.deleteQuietly(aria2cDir);
            aria2cDir.mkdirs();
            try {
                ZipUtils.unzip(aria2cLib, aria2cDir);
            } catch (Exception e) {
                FileUtils.deleteQuietly(aria2cDir);
                throw new YoutubeDLException("failed to initialize", e);
            }
            updateAria2c(appContext, aria2cSize);
        }
    }

    private boolean shouldUpdateAria2c(@NonNull Context appContext, @NonNull String version) {
        return !version.equals(SharedPrefsHelper.get(appContext, aria2cLibVersion));
    }

    private void updateAria2c(@NonNull Context appContext, @NonNull String version) {
        SharedPrefsHelper.update(appContext, aria2cLibVersion, version);
    }
}

File: aria2c/src/main/java/com/yausername/aria2c/Aria2c.kt
package com.yausername.aria2c

import android.content.Context
import com.yausername.youtubedl_android.YoutubeDLException
import com.yausername.youtubedl_common.SharedPrefsHelper
import com.yausername.youtubedl_common.utils.ZipUtils
import org.apache.commons.io.FileUtils
import java.io.File

object Aria2c {
    private var initialized = false
    private var binDir: File? = null

    @Synchronized
    fun init(appContext: Context) {
        if (initialized) return
        val baseDir = File(appContext.noBackupFilesDir, baseName)
        if (!baseDir.exists()) baseDir.mkdir()
        binDir = File(appContext.applicationInfo.nativeLibraryDir)
        val packagesDir = File(baseDir, packagesRoot)
        val aria2cDir = File(packagesDir, aria2cDirName)
        initAria2c(appContext, aria2cDir)
        initialized = true
    }

    private fun initAria2c(appContext: Context, aria2cDir: File) {
        val aria2cLib = File(binDir, aria2cLibName)
        if (!aria2cLib.exists()) {
            return
        }
        // using size of lib as version
        val aria2cSize = aria2cLib.length().toString()
        if (!aria2cDir.exists() || shouldUpdateAria2c(appContext, aria2cSize)) {
            FileUtils.deleteQuietly(aria2cDir)
            aria2cDir.mkdirs()
            try {
                ZipUtils.unzip(aria2cLib, aria2cDir)
            } catch (e: Exception) {
                FileUtils.deleteQuietly(aria2cDir)
                throw YoutubeDLException("failed to initialize", e)
            }
            updateAria2c(appContext, aria2cSize)
        }
    }

    private fun shouldUpdateAria2c(appContext: Context, version: String): Boolean {
        return version != SharedPrefsHelper.get(appContext, aria2cLibVersion)
    }

    private fun updateAria2c(appContext: Context, version: String) {
        SharedPrefsHelper.update(appContext, aria2cLibVersion, version)
    }

    private const val baseName = "youtubedl-android"
    private const val packagesRoot = "packages"
    private const val aria2cDirName = "aria2c"
    private const val aria2cLibName = "libaria2c.zip.so"
    private const val aria2cLibVersion = "aria2cLibVersion"

    @JvmStatic
    fun getInstance() = this
}